[REEF-492] A HDFS implementation of IFileSystem

markusweimer opened the pull request:
This adds an `IFileSystem` implementation using the `hdfs` command available on Hadoop. It also adds tests for it, which have been disabled as they only work if HDFS is available.

JIRA: [REEF-492](https://issues.apache.org/jira/browse/REEF-492)

markusweimer commented:
Rebased onto current `master`

markusweimer commented:
Rebased onto current `master`

markusweimer commented:
Rebased onto current `master`

afchung commented on lang/cs/Org.Apache.REEF.IO/FileSystem/Hadoop/HDFSCommandRunner.cs:
```diff
@@ -0,0 +1,172 @@
+            }
+            else
+            {
+                _hdfsCommandPath = Path.Combine(hadoopHome, "bin", "hdfs.cmd");
```
Might want to make "bin" and "hdfs.cmd" constants.

markusweimer replied:
Yes, will do.

afchung commented on lang/cs/Org.Apache.REEF.IO/FileSystem/Hadoop/HDFSCommandRunner.cs:
```diff
@@ -0,0 +1,172 @@
+                Logger.Log(Level.Info, "Trial {0} Timeout in {1} secs Executing: {2}", origTries - tries,
+                    _timeOutInMilliSeconds/1000,
+                    _hdfsCommandPath + " " + hdfsCommandLineArguments);
+                var startInfo = new ProcessStartInfo
```
Would it be better to use new Process and register the callbacks first instead starting with ProcessStartInfo and then registering the callbacks?

markusweimer replied:
Not sure I understand the question. Can you sketch out the alternative in code?

afchung replied:
var process = new Process();
process.callback += function();
process.Start();

afchung replied:
The way it's implemented here I'm not sure if there will be some sort of race condition, though I'm not that familiar with the process class.

afchung commented on lang/cs/Org.Apache.REEF.IO/FileSystem/Hadoop/HDFSCommandRunner.cs:
```diff
@@ -0,0 +1,172 @@
+                Logger.Log(Level.Warning, "Process has not exited after a timeout of {0} secs. So killing it",
+                    _timeOutInMilliSeconds);
+                process.Kill();
+            } while (--tries > 0);
```
Why not use a regular for loop?

markusweimer replied:
Uhm, because this made more sense to @shravanmn when he wrote the code when he wrote the code I ~~blatantly ripped off~~ used as generous inspiration for this? :-)

Seriously, though: I agree with the sentiment that a `for` loop would be easier to understand here and will change the code to use one.

jwang98052 commented on lang/cs/Org.Apache.REEF.IO.Tests/TestHadoopFileSystem.cs:
```diff
@@ -0,0 +1,138 @@
+    [Ignore] // These tests need to be run in an environment with HDFS installed.
+    public sealed class TestHadoopFileSystem
+    {
+        private readonly HadoopFileSystem _fileSystem =
```
Are you going to use singleton for HadoopFileSystem  in REEF? If yes, you might want  to put this code in  [ClassInitialize] method so that it will be only created once for all the tests in the class.

markusweimer replied:
Done.

jwang98052 commented on lang/cs/Org.Apache.REEF.IO/FileSystem/Hadoop/CommandResult.cs:
```diff
@@ -0,0 +1,51 @@
+    internal sealed class CommandResult
+    {
+        private readonly int _exitCode;
+        private readonly List<string> _stdErr;
```
Use IList

markusweimer replied:
Done.

jwang98052 commented on lang/cs/Org.Apache.REEF.IO/FileSystem/Hadoop/CommandResult.cs:
```diff
@@ -0,0 +1,51 @@
+    {
+        private readonly int _exitCode;
+        private readonly List<string> _stdErr;
+        private readonly List<string> _stdOut;
```
same

markusweimer replied:
Done.

jwang98052 commented on lang/cs/Org.Apache.REEF.IO/FileSystem/Hadoop/CommandResult.cs:
```diff
@@ -0,0 +1,51 @@
+            _exitCode = exitCode;
+        }
+
+        internal List<string> StdOut
```
ILIst would be better

markusweimer replied:
Done.

jwang98052 commented on lang/cs/Org.Apache.REEF.IO/FileSystem/Hadoop/CommandResult.cs:
```diff
@@ -0,0 +1,51 @@
+            get { return _stdOut; }
+        }
+
+        internal List<string> StdErr
```
same

markusweimer replied:
Done.

jwang98052 commented on lang/cs/Org.Apache.REEF.IO/FileSystem/Hadoop/HDFSCommandRunner.cs:
```diff
@@ -0,0 +1,194 @@
+            _numberOfRetries = numberOfRetries;
+            _timeOutInMilliSeconds = timeOutInMilliSeconds;
+
+            if (hadoopHome == HadoopHome.DefaultValue)
```
== works here if the client doesn't set it. What if the user set the that is the same a default value? Then you might want to use Equals.

markusweimer replied:
Good catch, thanks! Fixed.

jwang98052 commented on lang/cs/Org.Apache.REEF.IO/FileSystem/Hadoop/HDFSCommandRunner.cs:
```diff
@@ -0,0 +1,194 @@
+            }
+            else
+            {
+                _hdfsCommandPath = GetFullPathToHdfsCommand(hadoopHome);
```
Shall we do some validation in case the hadoop home that the user provided is invalid?

markusweimer replied:
Good idea. I added some code that tries the user-provided folder first. If that fails, it tries the environment variable.

jwang98052 commented on lang/cs/Org.Apache.REEF.IO/FileSystem/Hadoop/HadoopFileSystemConfiguration.cs:
```diff
@@ -0,0 +1,56 @@
+    /// <remarks>
+    /// This IFileSystem implementation is enormously slow as it spawns a new JVM per API call. Avoid if you have better means
+    /// of file system access.
+    /// Also, Stream-based operations are not supportted.
```
typo - supported

markusweimer replied:
fixed.

jwang98052 commented:
I completed my pass.

markusweimer commented:
@jwang98052 @afchung Thanks for the feedback! I will ping again when I'm done incorporating it.

markusweimer commented:
@jwang98052 @afchung I've addressed the feedback. Moving to a `for()` loop was a bit more involved than initially estimated, but this new structure should be nicer all around.

afchung commented on lang/cs/Org.Apache.REEF.IO.Tests/FileSystemTestUtilities.cs:
```diff
@@ -0,0 +1,78 @@
+        /// <param name="path2"></param>
+        /// <returns></returns>
+        internal static bool HaveSameContent(string path1, string path2)
+        {
```
Why is this comparing only a single byte?

afchung replied:
I see, because we're only writing a single byte in the test.

markusweimer replied:
Yes, exactly :-)

afchung commented on lang/cs/Org.Apache.REEF.IO/FileSystem/Hadoop/HDFSCommandRunner.cs:
```diff
@@ -0,0 +1,241 @@
+            _numberOfRetries = numberOfRetries;
+            _timeOutInMilliSeconds = timeOutInMilliSeconds;
+
+            if (!HadoopHome.DefaultValue.Equals(hadoopHome))
```
Windows paths are case insensitive.

afchung replied:
Or better, use suggestion provided here:
http://stackoverflow.com/questions/2281531/how-can-i-compare-directory-paths-in-c

markusweimer replied:
Good point, I will adapt that solution to our case.

afchung commented on lang/cs/Org.Apache.REEF.IO/FileSystem/Hadoop/HDFSCommandRunner.cs:
```diff
@@ -0,0 +1,241 @@
+            // Make sure we found the command.
+            if (!File.Exists(_hdfsCommandPath))
+            {
+                throw new Exception("HDFS command does not exist: " + _hdfsCommandPath);
```
Should we throw a more specific type of Exception?

markusweimer replied:
Good catch, replaced it with `System.IO.FileNotFoundException`

afchung commented:
Looks good to me other than a few small comments.

markusweimer commented:
@jwang98052 @afchung I've addressed all feedback.

afchung commented:
@markusweimer Looks good to me.

jwang98052 commented:
Looks good to me too.

jwang98052 commented:
I will test and merge it.
